examples/nlp/nl_image_search.py

```diff
@@ -0,0 +1,588 @@
+            (captions_similarity + images_similarity) / (2 * self.temperature)
+        )
+        # Compute the loss for the captions using crossentropy
+        captions_loss = keras.losses.categorical_crossentropy(
+            y_true=targets, y_pred=logits, from_logits=True
+        )
+        # Compute the loss for the images using crossentropy
+        images_loss = keras.losses.categorical_crossentropy(
+            y_true=tf.transpose(targets), y_pred=tf.transpose(logits), from_logits=True
+        )
+        # Return the mean of the loss over the batch.
+        return (captions_loss + images_loss) / 2
+
+    def train_step(self, features):
+        with tf.GradientTape() as tape:
+            # Forward pass
+            caption_embeddings, image_embeddings = self(features, training=True)
+            loss = self.compute_loss(caption_embeddings, image_embeddings)
+        # Backward pass
+        gradients = tape.gradient(loss, self.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))
+        # Monitor loss
+        self.loss_tracker.update_state(loss)
+        return {"loss": self.loss_tracker.result()}
+
+    def test_step(self, features):
+        caption_embeddings, image_embeddings = self(features, training=False)
+        loss = self.compute_loss(caption_embeddings, image_embeddings)
+        self.loss_tracker.update_state(loss)
+        return {"loss": self.loss_tracker.result()}
+
+
+"""
+## Train the dual encoder model
+
+In this experiment, we freeze the base encoders for text and images, and make only
+the projection head trainable.
+"""
+
+num_epochs = 5
+batch_size = 256
+
+vision_encoder = create_vision_encoder(
+    num_projection_layers=1, projection_dims=256, dropout_rate=0.1
+)
+text_encoder = create_text_encoder(
+    num_projection_layers=1, projection_dims=256, dropout_rate=0.1
+)
+dual_encoder = DualEncoder(text_encoder, vision_encoder, temperature=0.05)
+dual_encoder.compile(
+    optimizer=tfa.optimizers.AdamW(learning_rate=0.001, weight_decay=0.001)
+)
+
+"""
+Note that training the model with 60,000 image-caption pairs, with a batch size of 256,
+takes around 12 minutes per epoch using a V100 GPU accelerator. If 2 GPUs are available,
+the epoch takes around 8 minutes.
+"""
+
+print(f"Number of GPUs: {len(tf.config.list_physical_devices('GPU'))}")
+print(f"Number of examples (caption-image pairs): {train_example_count}")
+print(f"Batch size: {batch_size}")
+print(f"Steps per epoch: {int(np.ceil(train_example_count / batch_size))}")
+train_dataset = get_dataset(os.path.join(tfrecords_dir, "train-*.tfrecord"), batch_size)
+valid_dataset = get_dataset(os.path.join(tfrecords_dir, "valid-*.tfrecord"), batch_size)
+# Create a learning rate scheduler callback.
+reduce_lr = keras.callbacks.ReduceLROnPlateau(
+    monitor="val_loss", factor=0.2, patience=3
+)
+# Create an early stopping callback.
+early_stopping = tf.keras.callbacks.EarlyStopping(
```

Comment:
keras is already imported from tensorflow. For sake of consistency with other callback just a few lines above we can drop `tf.` part